test/spec/Garn/CodeGenSpec.hs

```diff
@@ -2,161 +2,372 @@
+                nixRaw`pkgs.nested-1.nested-2.some-pkg`,
+                "",
+              );
+            |]
+
+    testWritePkgFiles
+      "sanitizes package names"
+      ( ("a-package" ~> Derivation {description = Nothing, path = "pkgs.a-package"})
+          <> ("a/b@c" ~> Derivation {description = Nothing, path = "pkgs.a/b@c"})
+          <> ("catch" ~> Derivation {description = Nothing, path = "pkgs.catch"})
+          <> ("instanceOf" ~> Derivation {description = Nothing, path = "pkgs.instanceOf"})
+          <> ("super" ~> Derivation {description = Nothing, path = "pkgs.super"})
+          <> ("void" ~> Derivation {description = Nothing, path = "pkgs.void"})
+      )
+      $ \readFile -> do
+        readFile "mod.ts"
+          `shouldReturn` unindent
+            [i|
+              import { mkPackage } from "path/to/garn/root/package.ts";
+              import { nixRaw } from "path/to/garn/root/nix.ts";
+
+              export const a_b_c = mkPackage(
+                nixRaw`pkgs.a/b@c`,
+                "",
+              );
+
+              export const a_package = mkPackage(
+                nixRaw`pkgs.a-package`,
+                "",
+              );
+
+              export const catch_ = mkPackage(
+                nixRaw`pkgs.catch`,
+                "",
+              );
+
+              export const instanceOf_ = mkPackage(
+                nixRaw`pkgs.instanceOf`,
+                "",
+              );
+
+              export const super_ = mkPackage(
+                nixRaw`pkgs.super`,
+                "",
+              );
+
+              export const void_ = mkPackage(
+                nixRaw`pkgs.void`,
+                "",
+              );
+            |]
+    testWritePkgFiles
+      "removes conflicts due to sanitization"
+      ( ("a-package" ~> Derivation {description = Nothing, path = "pkgs.a-package"})
+          <> ("a_package" ~> Derivation {description = Nothing, path = "pkgs.a_package"})
+          <> ("a/package" ~> Derivation {description = Nothing, path = "pkgs.a/package"})
+      )
+      $ \readFile -> do
+        readFile "mod.ts"
+          `shouldReturn` unindent
+            [i|
+              import { mkPackage } from "path/to/garn/root/package.ts";
+              import { nixRaw } from "path/to/garn/root/nix.ts";
+
+              export const a_package = mkPackage(
+                nixRaw`pkgs.a_package`,
+                "",
+              );
+            |]
+
+testScanPackages :: String -> String -> String -> Map String PkgInfo -> SpecWith ()
+testScanPackages description pkgs pkgSpec expectedOutput =
+  it description $ do
+    withModifiedEnvironment [("NIX_CONFIG", "experimental-features =")] $ do
+      system <- currentSystem
+      actualOutput <- scanPackages system pkgs pkgSpec
+      actualOutput `shouldBe` expectedOutput
+
+testWritePkgFiles :: String -> Map String PkgInfo -> ((FilePath -> IO String) -> IO ()) -> SpecWith ()
+testWritePkgFiles description pkgInfo runTest =
```

Comment:
Before we called `deno check` to typecheck the files generated in these tests. Did you remove that intentionally? Should we get that back for the tests that use `testWritePkgFiles`?

Reply:
This wasn't entirely intentional, but I assumed on CI we were typechecking the codegen files. It looks like we do call `test-ts` after `codegen` on GitHub, but at the moment this doesn't typecheck. I think maybe we should just bring over the change on #409 of adding `--check` to `test-ts` recipe. What do you think? Is there an advantage to also running typecheck in the Haskell tests?

Reply:
> [...] What do you think? Is there an advantage to also running typecheck in the Haskell tests?

I think the advantage was mostly during development, that it was nice to TDD on smaller examples and have `deno` yell at you when you weren't generating valid typescript. So I thought it was nice to have. But you were also the last person to work on this code, and if you didn't miss it, maybe it's not that useful? Up to you!

And yes, please add `--check` to `test-ts`! That makes sense in any case.

Reply:
> And yes, please add `--check` to `test-ts`! That makes sense in any case.

Never mind, `typescript-check` is a post-dependency of the `codegen` recipe: https://github.com/garnix-io/garn/blob/f80b691b422d7cc84a57bb2693ae9c25071c44b7/justfile#L158. So we do typecheck on CI and everytime we run `just codegen`.